server/commands/write.go

```diff
@@ -104,3 +109,73 @@ func handleError(err error) error {
 
 	return serverErrors.HandleError("", err)
 }
+
+func validateHasDirectRelationship(tupleUserset *openfgapb.Userset, tk *openfgapb.TupleKey) error {
+	indirectWriteErrorReason := "Attempting to write directly to an indirect only relationship"
+	switch usType := tupleUserset.Userset.(type) {
+	case *openfgapb.Userset_Intersection:
+		if !isDirectIntersection(usType, tk) {
+			return serverErrors.HandleTupleValidateError(&tupleUtils.IndirectWriteError{Reason: indirectWriteErrorReason, TupleKey: tk})
+		}
+	case *openfgapb.Userset_Union:
+		if !isDirectUnion(usType, tk) {
+			return serverErrors.HandleTupleValidateError(&tupleUtils.IndirectWriteError{Reason: indirectWriteErrorReason, TupleKey: tk})
+		}
+	case *openfgapb.Userset_Difference:
+		if !isDirectDifference(usType, tk) {
+			return serverErrors.HandleTupleValidateError(&tupleUtils.IndirectWriteError{Reason: indirectWriteErrorReason, TupleKey: tk})
+		}
+	case *openfgapb.Userset_ComputedUserset:
+		// if Userset.type is a ComputedUserset then we know it can't be direct
+		return serverErrors.HandleTupleValidateError(&tupleUtils.IndirectWriteError{Reason: indirectWriteErrorReason, TupleKey: tk})
+	default:
+		return nil
+	}
+	return nil
+}
+
+func isDirectIntersection(nodes *openfgapb.Userset_Intersection, tk *openfgapb.TupleKey) bool {
+	isDirect := false
```

Comment:
I think we can get rid of this variable entirely. On line 142 return true, and on line 149 return false. I think that would be even easier to read.

Reply:
Another great point -- I kept it in since it was there already, but we've covered all return possibilities and the code is simpler 👍

Reply:
The latest commit has the cleaned up version!